File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <bits/stdc++.h>
using namespace std;

string braces;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    
	while (true) {
		getline(cin, braces);

		if (braces == ".")
			break;

		stack<char> stk;
		bool flag = true;

		for (char c : braces) {
			if (c != '(' && c != ')' && c != '[' && c != ']')
				continue;

			if (c == '(' || c == '[') {
				stk.push(c);
			}
			else if (c == ')' && stk.size() && stk.top() == '(') {
				stk.pop();
			}
			else if (c == ']' && stk.size() && stk.top() == '[') {
				stk.pop();
			}
			else {
				flag = false;
                break;
			}
		}

		if (stk.size()) flag = false;

		if (flag) cout << "yes\n"; else cout << "no\n";
	}
	return 0;
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
input = sys.stdin.readline

while True:
    text = input().rstrip()

    if text == ".":
        break

    stk = []

    flag = True

    for c in text:
        if c in {"(", "["}:
            stk.append(c)

        elif c == ")":
            if stk and stk[-1] == "(":
                stk.pop()
            else:
                flag = False
                break

        elif c == "]":
            if stk and stk[-1] == "[":
                stk.pop()
            else:
                flag = False
                break

    if stk:
        flag = False

    print("yes" if flag else "no")
